### PR TITLE
Removed filter for attendee e-mail so that you can select an account …

### LINKED ISSUE #492: trying to create admins when they are set to register themselves and no email was provided fails in a confusing way
In ubersystem, when you want to add someone as an admin account, you have to first add them into the system as a regular attendee.  Once they are a regular attendee, you can promote them to a full administrator.

The bug here is that if you add an attendee without an email address, they (confusingly) don't show up as an option to add them as an administrator.

This is because the system won't let you add an admin without an email address, because it needs to send an email to that person to get them a link they can use to login.

How we should fix it:
1) Make people without emails show up in the list by modifying the database query
2) When we try and add anyone as an admin, check to make sure they have a valid email, and if they don't, display an error message saying 'this attendee needs a valid email, please fix'

Reproduce steps:

Go to People -> Attendees:
image

Click top-level 'Add Attendee', fill in the following:
1) First name: Cool
2) Last name: Guy
3) check the checkbox for 'register themselves'
4) leave email field blank

image

Click top-level 'Accounts' link:
Click the dropdown that shows attendee's names:
image

EXPECTED BEHAVIOR: Cool Guy is shown in the list.  If you select 'Create Account' it should fail with an error message that says 'Cool Guy' doesn't have a valid email address, please add one first.
ACTUAL BEHAVIOR: Cool Guy does not show up in the list.

Tech notes
===

This line here is the DB query:
```python
Attendee.objects.exclude(email='')
```

To return an error message to the user, you need to pass 'message' to the template engine something like this:

```python
return {'message': "can't create an admin account for user because attendee doesn't have a valid email address set, please set one first" }
```

-------
(low priority cause this is kind of an edge case, but documenting nonetheless)

### PR DESCRIPTION
**Pull request by Lusarien**
…without an email, but with throw an error if you attempt to make the account an admin.

Fixes #492

### CODE REVIEW

File: uber/model_checks.py [hunk 1]
```diff
@@ -23,6 +23,8 @@ def duplicate_admin(account):
         with Session() as session:
             if session.query(AdminAccount).filter_by(attendee_id=account.attendee_id).all():
                 return 'That attendee already has an admin account'
+            elif session.query(Attendee).filter_by(id=account.attendee_id).first().email == '':
+                return "Attendee doesn't have a valid email set"
```
Comment (binary1230) [inline uber/model_checks.py]: I think it might be a good idea to put this in a new validator that is named 'check_valid_email_exists', or something like that. (since this one is named duplicate_admin, and this check isn't related to that)

File: uber/model_checks.py [hunk 2]
```diff
@@ -24,6 +24,13 @@ def duplicate_admin(account):
             if session.query(AdminAccount).filter_by(attendee_id=account.attendee_id).all():
                 return 'That attendee already has an admin account'
 
+@validation.AdminAccount
+def duplicate_admin(account):
+    if account.is_new:
+        with Session() as session:
+            if session.query(Attendee).filter_by(id=account.attendee_id).first().email == '':
```
Comment (EliAndrewC) [inline uber/model_checks.py]: There's a convenience method you could use here; you can just say ``session.attendee(account.attendee_id)`` to get the Attendee with that id.  This exists for all of our models (e.g. you could say ``session.job(some_id)`` or ``session.event(another_id)``).

So in this case you should probably rewrite this if statement to just say

```python
if not session.attendee(account.attendee_id).email:
```

(In this case I did ``if not`` instead of ``== ''`` but that's mostly a stylistic preference, and you could do it either way.)

**Comment by EliAndrewC** [comment]
Looks great!  Other than the one suggestion I made (which is just a suggestion for a shorter way to write one of the working lines of code you already implemented), the one thing you'll need to fix is that there's a PEP-8 violation which is causing Travis to throw an error:
https://travis-ci.org/magfest/ubersystem/builds/66431670

In case it's not clear what this means, we run our code against a formatting checker (https://pypi.python.org/pypi/pep8) so from your dev environment you can just say something like ``pep8 plugins/uber/uber`` and it will run a bunch of style checks to make sure we're following generally accepted coding conventions.  In this case, you've defined a function without having two blank lines before it.  If you add an extra line and them commit and push to this branch, Travis will re-run automatically and should pass, at which point we can merge this in.

File: uber/model_checks.py [hunk 3]
```diff
@@ -25,6 +25,14 @@ def duplicate_admin(account):
                 return 'That attendee already has an admin account'
 
 
+@validation.AdminAccount
+def duplicate_admin(account):
```
Comment (kitsuta) [inline uber/model_checks.py]: Since there's already a function called `duplicate_admin`, and I'm pretty sure Python won't let you overload functions, this wouldn't work. You'll want to change your function's name to better reflect what it does, anyway.

Comment (EliAndrewC) [inline uber/model_checks.py]: Good catch!  Yeah, this should definitely be changed to say something like ``def has_email_addr`` or something.

**Comment by kitsuta** [comment]
Just to expand a bit on the line comment:

Python does not allow method overloading because it looks functions up purely by name (it doesn't consider the parameters to be part of the function's internal "ID"). When there are two methods with the same name in the same class, the latter just overwrites the former. So this function would work fine, but the original `duplicate_admin` method would never be run, which we definitely don't want because that means we'd stop actually checking for duplicate admin accounts.

Plus, we generally want method names to be self-explanatory, so we don't have to think too hard to remember what they do. :)

**Comment by Lusarien** [comment]
Oops, thanks for the catch, meant to re-name that I will fix that right now

File: uber/site_sections/accounts.py [hunk 4]
```diff
@@ -20,8 +20,7 @@ def index(self, session, message=''):
                                .order_by(Attendee.first_name, Attendee.last_name).all(),
             'all_attendees': sorted([
                 (id, '{} - {}{}'.format(name.title(), c.BADGES[badge_type], ' #{}'.format(badge_num) if badge_num else ''))
-                for id, name, badge_type, badge_num in session.query(Attendee.id, Attendee.full_name, Attendee.badge_type, Attendee.badge_num)
-                                                              .filter(Attendee.email != '').all()
+                for id, name, badge_type, badge_num in session.query(Attendee.id, Attendee.full_name, Attendee.badge_type, Attendee.badge_num).all()
```
Comment (kitsuta) [inline uber/site_sections/accounts.py]: I'm a little confused about the logic behind this. Maybe there's an issue I missed that discussed the reasoning behind doing this?

It makes sense that we never want to create admin accounts for people without email addresses - we need to email them their password, for example - but why do we include those people in the list of the Accounts page? The user would assume that anyone in the list can have an account, then be disappointed when they get an error. It's a somewhat cumbersome way to go about it - you'd hunt for the right name in the list, have the person enter their password, and _then_ find out that they can't have an account until you go to the Attendees page, find them, set their email, and then go back and find their name in the list of attendees under the Accounts page.

That doesn't prevent us from merging this PR, but it's definitely something to think about. Maybe a quick fix could be to display an email field that you can enter the email in while creating the admin account?

Comment (EliAndrewC) [inline uber/site_sections/accounts.py]: The reason this was proposed is that we had a bunch of confused admins this year who kept trying to create another admin at regdesk, but they couldn't figure out why the attendee's name wasn't on the list.  The suggestion at the time was, "Let's put all of the names on the list, so they can select the person, and then they get told that the person doesn't have an email address so that people aren't confused."

There are probably better solutions, but this seems like an okay quick-and-easy thing to prevent that from happening.

Although come to think of it, we should probably prevent people without first/last names from showing up here.  So maybe include a ``.filter(Attendee.first_name != '')`` in the query, since otherwise we'll have tons and tons of unclaimed group badges showing up in the dropdown, which seems bad.

**Comment by binary1230** [comment]
So what ended up happening was,

I was trying to add someone as an admin that had previously been added to the system without an email being put in, which is allowed when you do 'attendee needs to register themselves'

When trying to add this person as an Attendee, it wasn't showing up in the list which was super-confusing.  I tried a bunch of stuff and it seemed like something was broken because why wouldn't this person be in the list.

Finally, I copied the DB and looked at this locally, and found out it doesn't show people without emails in the list.

So, basically, I was just thinking it would be better behavior if the system at least had a way to tell you what was wrong with that particular attendee, rather than silently failing.  

This was a pretty low prioritiy, but I had Brian work on it just because it would be a good first bug to mess around with to cut his teeth on getting everything setup and working with the code.

I'm down for further UX improvements, I just figured doing the change this way was 'less broken' than before, and a quick, neat fix.

**Comment by binary1230** [comment]
btw hahah I can also take credit for the duplicately named function because to demonstrate code review, I had @Lusarien quickly check in a copy of that change so it could be seen in this PR.  so, uh... woops :)  welcome onboard dude :+1:

File: uber/site_sections/accounts.py (hunk 4 above)
Comment (kitsuta) [inline uber/site_sections/accounts.py]: Yeah, having a bajillion unclaimed badges is my main concern for this. @Lusarien can you add that check to the query?

Further UX improvements can be added as issues so we can address them later - that's more of a front-end concern anyway.

**Comment by kitsuta** [comment]
Looks good with the new changes!
